### PR TITLE
chore: setting up dependency factories

### PR DESCRIPTION
**Pull request by ScriptedAlchemy**

### CODE REVIEW

File: src/webpack-core/ContainerPlugin.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+  needBuild(context, callback) {
+    return callback(null, !this.buildMeta);
+  }
+
+  build(options, compilation, resolver, fs, callback) {
+    this.buildMeta = {};
+    this.buildInfo = {
+      strict: true,
+      builtTime: Date.now(),
```
Comment (sokra) [inline src/webpack-core/ContainerPlugin.js]: You don't need that

Other modules use a build time to determine if the module need to be rebuild. This module never need to be rebuild.

File: src/webpack-core/ContainerPlugin.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    for (const dep of this.expose) {
+      const block = new AsyncDependenciesBlock(
+        undefined,
+        dep.loc,
+        dep.userRequest
+      );
+      block.addDependency(dep);
+      this.addBlock(block);
+
+      this.buildInfo.expose.set(dep.exposedName, block);
```
Comment (sokra) [inline src/webpack-core/ContainerPlugin.js]: `buildInfo` is serialized during persistent caching. While this is working, it's unneeded.

You can iterate all blocks and read the `exposedName` from the first dependency.

File: src/webpack-core/ContainerPlugin.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    const getters = [];
+
+    for (const [name, block] of this.buildInfo.expose) {
+      getters.push(
+        `case "${name}":\nreturn ${runtimeTemplate.blockPromise({
+          block,
+          message: `${name}`, // TODO: Should we use the request here?
+          chunkGraph,
+          runtimeRequirements
+        })};`
```
Comment (sokra) [inline src/webpack-core/ContainerPlugin.js]: You also want to add a `.then(() => () => __webpack_require__(123))` to return a module factory function.

File: src/webpack-core/ContainerPlugin.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      exports["get"] = function get(module) {
+        switch(module) {
+            ${getters.join("\n")}
+          default:
+            return Promise.resolve().then(() => { throw new Error('Module ' + module + ' does not exist!'); });
+        }
+      }
+    `);
+
+    this.clearDependenciesAndBlocks();
```
[the remark this answers is not in the document]
Comment (maraisr) [reply src/webpack-core/ContainerPlugin.js]: Ah yeah, thanks. Wasnt sure. When i had `this.addDepedency` they appeared concatenated with this module. So probably a remittance of trying to remove them.

File: src/webpack-core/ContainerPlugin.js (hunk 1 above)
Comment (maraisr) [reply src/webpack-core/ContainerPlugin.js]: Ah perfect! Thought that had something to do with stats or something. Removing! :100:

File: src/webpack-core/ContainerPlugin.js (hunk 2 above)
Comment (maraisr) [reply src/webpack-core/ContainerPlugin.js]: Had that initially, just wasn't 100% sure that - that single dependency was always going to be that. Will use that.

File: src/webpack-core/ContainerPlugin.js (hunk 3 above)
Comment (maraisr) [reply src/webpack-core/ContainerPlugin.js]: hang on, so should I assign that `runtimeTemplate.blockPromise` to a local variable or something? And go: `const x = runtimeTemplate.blockPromise(...); "return ${x}.then(() => ${x})"`

File: src/webpack-core/ContainerPlugin.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        runtimeRequirements
+      });
+
+      getters.push(
+        `case "${name}":\nreturn ${require_statement}.then(() => ${RuntimeGlobals.require}(${moduleId}));`
+      );
+    }
+
+    source.add(`
+      exports["get"] = function get(module) {
```
Comment (maraisr) [inline src/webpack-core/ContainerPlugin.js]: @sokra is there webpack internals I can use to generate the `__webpack_exports__` block like harmony does?

Screen Shot 2020-02-14 at 23 47 14

**Comment by ScriptedAlchemy** [comment]
Merging this into the core branch. So it’s easier for us to review and request reviews.

**Pull request merged**